File: app/src/main/java/com/example/habitpalette/ui/home/PastHabitRecyclerAdapter.java
package com.example.habitpalette.ui.home;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.habitpalette.ui.common.HabitScoreProgressView;
import com.example.habitpalette.R;

import java.util.ArrayList;

public class PastHabitRecyclerAdapter extends RecyclerView.Adapter<PastHabitRecyclerAdapter.ViewHolder> {

    private ArrayList<PastHabitRecyclerItem> mData;

    PastHabitRecyclerAdapter(ArrayList<PastHabitRecyclerItem> list) {
        mData = list;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        View view = inflater.inflate(R.layout.item_past_habit, parent, false);
        PastHabitRecyclerAdapter.ViewHolder vh = new PastHabitRecyclerAdapter.ViewHolder(view);

        return vh;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        PastHabitRecyclerItem item = mData.get(position);
        int score = (int) item.getHabitScore()*20;
        holder.habitTitle.setText(item.getHabitTitle());
        holder.habitScoreTextView.setText("평균점수: "+String.valueOf(item.getHabitScore()));
        holder.habitPeriod.setText(item.getHabitStartDate()+"~"+item.getHabitEndDate());
        holder.habitScoreView.setPercent(score);
        holder.habitScoreView.setText(Float.toString(item.getHabitScore()));
        holder.habitScoreView.setFgColorStart(item.getHabitColor());
        holder.habitScoreView.setFgColorEnd(0xff65b8b5);
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView habitTitle;
        TextView habitPeriod;
        TextView habitScoreTextView;
        HabitScoreProgressView habitScoreView;

        ViewHolder(View itemView) {
            super(itemView);

            habitTitle = itemView.findViewById(R.id.text_habit_title);
            habitPeriod = itemView.findViewById(R.id.text_habit_period);
            habitScoreTextView=itemView.findViewById(R.id.text_habit_score);
            habitScoreView = itemView.findViewById(R.id.progress_view_past_habit_score);
        }
    }
}

File: app/src/main/java/com/example/habitpalette/ui/home/PastHabitRecyclerAdapter.kt
package com.example.habitpalette.ui.home

import android.annotation.SuppressLint
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import com.example.habitpalette.data.model.Habit
import com.example.habitpalette.databinding.ItemPastHabitBinding

class PastHabitRecyclerAdapter:
    ListAdapter<Habit, PastHabitRecyclerAdapter.PastHabitViewHolder>(PastHabitComparator()){

    private lateinit var itemBinding: ItemPastHabitBinding

    inner class PastHabitViewHolder(private val itemBinding: ItemPastHabitBinding): RecyclerView.ViewHolder(itemBinding.root){
        private var habit:Habit? = null

        // item 클릭해서 CalendarActivity로 intent

        @SuppressLint("SetTextI18n")
        fun bind(data: Habit) {
            this.habit = data
            itemBinding.tvTitle.text = data.title
            itemBinding.tvScore.text = data.score.toString()
            itemBinding.tvPeriod.text = data.start_date.toString() + "~" + data.end_date.toString()
            itemBinding.pvPastProgress.percent = 50f
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PastHabitViewHolder {
        itemBinding = ItemPastHabitBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return PastHabitViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: PastHabitViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    // Room을 사용할 경우 diffUtil로 notifyFataSetChanged 대체 가능
    class PastHabitComparator: DiffUtil.ItemCallback<Habit>(){
        override fun areItemsTheSame(oldItem: Habit, newItem: Habit): Boolean {
            return oldItem === newItem
        }

        override fun areContentsTheSame(oldItem: Habit, newItem: Habit): Boolean {
            return oldItem.id == newItem.id
        }
    }
}